src/CallHandler.tsx

```diff
@@ -0,0 +1,482 @@
+        // audio clips in to play.
+        if (navigator.mediaSession) {
+            navigator.mediaSession.setActionHandler('play', function() {});
+            navigator.mediaSession.setActionHandler('pause', function() {});
+            navigator.mediaSession.setActionHandler('seekbackward', function() {});
+            navigator.mediaSession.setActionHandler('seekforward', function() {});
+            navigator.mediaSession.setActionHandler('previoustrack', function() {});
+            navigator.mediaSession.setActionHandler('nexttrack', function() {});
+        }
+    }
+
+    getCallForRoom(roomId: string) {
+        return this.calls[roomId] || null;
+    }
+
+    getAnyActiveCall() {
+        const roomsWithCalls = Object.keys(this.calls);
+        for (let i = 0; i < roomsWithCalls.length; i++) {
+            if (this.calls[roomsWithCalls[i]] &&
+                    this.calls[roomsWithCalls[i]].call_state !== "ended") {
+                return this.calls[roomsWithCalls[i]];
+            }
+        }
+        return null;
+    }
+
+    play(audioId) {
+        // TODO: Attach an invisible element for this instead
+        // which listens?
+        const audio = document.getElementById(audioId) as HTMLMediaElement;
+        if (audio) {
+            const playAudio = async () => {
+                try {
+                    // This still causes the chrome debugger to break on promise rejection if
+                    // the promise is rejected, even though we're catching the exception.
+                    await audio.play();
+                } catch (e) {
+                    // This is usually because the user hasn't interacted with the document,
+                    // or chrome doesn't think so and is denying the request. Not sure what
+                    // we can really do here...
+                    // https://github.com/vector-im/element-web/issues/7657
+                    console.log("Unable to play audio clip", e);
+                }
+            };
+            if (this.audioPromises[audioId]) {
+                this.audioPromises[audioId] = this.audioPromises[audioId].then(() => {
+                    audio.load();
+                    return playAudio();
+                });
+            } else {
+                this.audioPromises[audioId] = playAudio();
+            }
+        }
+    }
+
+    pause(audioId) {
+        // TODO: Attach an invisible element for this instead
+        // which listens?
+        const audio = document.getElementById(audioId) as HTMLMediaElement;
+        if (audio) {
+            if (this.audioPromises[audioId]) {
+                this.audioPromises[audioId] = this.audioPromises[audioId].then(() => audio.pause());
+            } else {
+                // pause doesn't actually return a promise, but might as well do this for symmetry with play();
+                this.audioPromises[audioId] = audio.pause();
+            }
+        }
+    }
+
+    private setCallListeners(call) {
```

Comment:
what type is this?

Reply:
it's a call object from js-sdk which I've just given an alias for any until we ts-ify the js-sdk bits